### ISSUE TITLE
k8s 1.19.4 kubelet stops presenting kubelet_volume_stats_used_bytes metrics

### ISSUE THREAD

**Issue by briend** [kubernetes/kubernetes#97138]
**What happened**:

After upgrading to k8s 1.19.4, kubelet metrics for `kubelet_volume_stats_used_bytes` and other `kubelet_volume_*` metrics are missing.  All other kubelet metrics seem to be fine.

**What you expected to happen**:

kubelet should continue providing these metrics

**How to reproduce it (as minimally and precisely as possible)**:

Perform a normal upgrade of k8s to 1.19.x and then examine the kubelet metrics endpoint on a node with Bound PVs:

```
curl -sS -k --header "Authorization: Bearer $TOKEN" https://node:10250/metrics | grep kubelet_volume
```

**Anything else we need to know?**:

We are using local volume PVs similar to this, managed by local-volume-provisioner:

```
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: local-volume-provisioner-nodex-4aa7e2ee-ef91-49fb-9b8a-9b5d0ae117f1
  creationTimestamp: "2019-11-26T23:45:12Z"
  finalizers:
  - kubernetes.io/pv-protection
  labels:
    kubernetes.io/hostname: node1
  name: local-pv-f92a7d5
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: node1
    uid: 4aa7e2ee-ef91-49fb-9b8a-9b5d0ae117f1
  resourceVersion: "17287578"
  selfLink: /api/v1/persistentvolumes/local-pv-f92a7d5
  uid: 389cf0ce-b91e-49ab-9c34-2db63b0bc956
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 48Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: vol0-test-mon-all-0
    namespace: noaa
    resourceVersion: "17285315"
    uid: 38eed1b9-8075-49fe-9002-06d7fa57809d
  local:
    path: /mnt/local-50g/sdc1
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-50g
  volumeMode: Filesystem
status:
  phase: Bound
```

I created a new k8s `1.18.4` cluster in a vagrant environment and upgraded it to k8s `1.19.4` and I could not reproduce this.  However, it has  happened to three of our dev/testing k8s clusters and we haven't found any significant difference in the configuration yet.  

Downgrading the kubelet binaries/pkgs alone to `1.18.x` resolves the issue (the metrics reappear), and as soon as you upgrade the binaries to any `1.19.x` release the metrics disappear again.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:17:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:09:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
local physical nodes with local storage
- OS (e.g: `cat /etc/os-release`):
both Bionic and Focal releases of Ubuntu

```
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
```

- Kernel (e.g. `uname -a`):
```
Linux 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
kubeadm v1.19.4

**Comment by pacoxu** [comment]
After #84907, kubelet 1.17 mark volume stats metrics as ALPHA.

> Following metrics from kubelet are now marked as with the ALPHA stability level:
> kubelet_container_log_filesystem_used_bytes
> kubelet_volume_stats_capacity_bytes
> kubelet_volume_stats_available_bytes
> kubelet_volume_stats_used_bytes
> kubelet_volume_stats_inodes
> kubelet_volume_stats_inodes_free
> kubelet_volume_stats_inodes_used
> plugin_manager_total_plugins
> volume_manager_total_volumes

I can reproduce it in my cluster of 1.19.3. It works in 1.18.5.

**Comment by mkhpalm** [comment]
For us they stop working after updating kubelet to v1.19.x (all versions) and then start working again if you update to v1.20.0. That said, if you join a new node with v1.19.x the metrics work. It only seems to affect upgraded nodes.

**Comment by pacoxu** [comment]
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/metrics/collectors/volume_stats.go#L91-L126

These metrics are added only if pods have some new pvc volume stats.
 
kubele_metrics_* metrics will  ignore if no PVC reference or already collected


After I create a pod using nfs, the metrics appears in my kubelet.

**Comment by pacoxu** [comment]
After restarting kubelet, the metrics is not found.
```
[root@daocloud-master ~]#  curl -k --key /etc/kubernetes/pki/apiserver-kubelet-client.key --cert /etc/kubernetes/pki/apiserver-kubelet-client.crt  --sa /etc/kubernetes/pki/ca.crt https://127.0.0.1:10250/metrics  |  grep volume_sta
curl: (3) <url> malformed
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 82715    0 82715    0     0  1878k      0 --:--:-- --:--:-- --:--:-- 1878k

```
After new nfs volume and pod are created, metrics goes correctly.

```
[root@daocloud-master ~]# kubectl create -f nfs.yaml
replicationcontroller/nfs-busybox-1 created
persistentvolumeclaim/nfs1 created
persistentvolume/nfs1 created

[root@daocloud-master ~]#  curl -k --key /etc/kubernetes/pki/apiserver-kubelet-client.key --cert /etc/kubernetes/pki/apiserver-kubelet-client.crt  --sa /etc/kubernetes/pki/ca.crt https://127.0.0.1:10250/metrics  |  grep volume_sta
curl: (3) <url> malformed
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP kubelet_volume_stats_available_bytes [ALPHA] Number of available bytes in the volume
# TYPE kubelet_volume_stats_available_bytes gauge
kubelet_volume_stats_available_bytes{namespace="default",persistentvolumeclaim="nfs1"} 8.8628789248e+10
# HELP kubelet_volume_stats_capacity_bytes [ALPHA] Capacity in bytes of the volume
# TYPE kubelet_volume_stats_capacity_bytes gauge
kubelet_volume_stats_capacity_bytes{namespace="default",persistentvolumeclaim="nfs1"} 1.051721728e+11
# HELP kubelet_volume_stats_inodes [ALPHA] Maximum number of inodes in the volume
# TYPE kubelet_volume_stats_inodes gauge
kubelet_volume_stats_inodes{namespace="default",persistentvolumeclaim="nfs1"} 5.1378176e+07
# HELP kubelet_volume_stats_inodes_free [ALPHA] Number of free inodes in the volume
# TYPE kubelet_volume_stats_inodes_free gauge
kubelet_volume_stats_inodes_free{namespace="default",persistentvolumeclaim="nfs1"} 5.1280932e+07
# HELP kubelet_volume_stats_inodes_used [ALPHA] Number of used inodes in the volume
# TYPE kubelet_volume_stats_inodes_used gauge
kubelet_volume_stats_inodes_used{namespace="default",persistentvolumeclaim="nfs1"} 97244
# HELP kubelet_volume_stats_used_bytes [ALPHA] Number of used bytes in the volume
# TYPE kubelet_volume_stats_used_bytes gauge
kubelet_volume_stats_used_bytes{namespace="default",persistentvolumeclaim="nfs1"} 1.6543383552e+10
100 84085    0 84085    0     0  2105k      0 --:--:-- --:--:-- --:--:-- 2105k


```

**Comment by mkhpalm** [comment]
Looks like this starts working again for us with kubelet 1.19.5 and 1.19.6 on bionic. It did not work on kubelet versions 1.19.0->1.19.4

**Comment by ehashman** [comment]
Is this bug still present on 1.20/master? I don't see anything in the 1.19.5 release notes suggesting a bug was fixed. Indeed, the file @pacoxu linked above hasn't been changed since 2019.

**Comment by pacoxu** [comment]
> Is this bug still present on 1.20/master? I don't see anything in the 1.19.5 release notes suggesting a bug was fixed. Indeed, the file @pacoxu linked above hasn't been changed since 2019.

After reading the code, I think the metrics should appear only if there're volume mounts in that node.
I cannot reproduce it in 1.19.0-1.19.4.

> Looks like this starts working again for us with kubelet 1.19.5 and 1.19.6 on bionic. It did not work on kubelet versions 1.19.0->1.19.4

The problem seems to be solved in 1.19.5.  
https://github.com/kubernetes/kubernetes/compare/v1.19.4...v1.19.5

I suspect it is fixed in Cadvisor or in #96477

1. github.com/google/cadvisor v0.37.0 -> v0.37.2 #96695  & #96849
2.  #96477 is also change volume metrics file in

https://github.com/kubernetes/kubernetes/blob/v1.19.4/pkg/volume/util/metrics.go#L117-L122
changed to 
https://github.com/kubernetes/kubernetes/blob/v1.19.4/pkg/volume/util/metrics.go#L119-L129

**Comment by ehashman** [comment]
Great! Glad it was fixed.
